package/metrics_daemon/metrics_daemon/src/metrics_daemon.c

```diff
@@ -0,0 +1,375 @@
+                                                       // of folder
+      json_object_object_add(jobj_lc_node, bname, jnewobj);
+    }
+    break;
+  }
+
+  case FTW_F: {
+    FILE *fp = fopen(fpath, "r"); // read file to get the data
+    char buf[64];
+    long file_len_long = sb->st_size - 1;
+    int file_len = (int)(sb->st_size - 1);
+    double number = 0.0;
+    if (file_len < 64 && file_len_long > 0) {
+      fread(&buf[0], (unsigned int)file_len, 1, fp);
+      number = atof(&buf[0]);
+    } else {
+      file_len = 0; // if file is empty, set the str length to zero
+    }
+    fclose(fp);
+    bool file_entry_exist = false;
+    json_object_object_foreach(jobj_lc_node, key, val)
+    {
+      if (strcmp(bname, key) == 0) {
+        file_entry_exist = true;
+        if (file_len == 0) {
+          json_object_put(val);
+          json_object_object_add(jobj_lc_node, bname, NULL);
+        } else {
+          json_object_set_double(val, number); // if file exist, update
+        }
+
+        break;
+      }
+    }
+    if (!file_entry_exist) // add a new node if it doesn't exist
+    {
+      if (file_len == 0)
+        json_object_object_add(jobj_lc_node, bname, NULL);
+      else
+        json_object_object_add(jobj_lc_node, bname, json_object_new_double(number));
+    }
+    break;
+  }
+  default: {
+    piksi_log(LOG_ERR,
+              "FTW: unexpected file type\n"); // For now, only support file and forlder
+    free(bname); // free the file name allocated under loop_through_folder_name()
+    return -1;
+  }
+  }
+
+  free(bname); // free the file name allocated under loop_through_folder_name()
+  return 0;
+}
+
+/**
+ * @brief function that write json to file
+ *
+ * Write messaging metrics to file
+ */
+static void write_metrics_to_file()
+{
+  // Walk dir for metrics
+  first_folder = true;
+  if (ftw(metrics_path, handle_walk_path, 20) == -1) {
+    return;
+  }
+  if (enable_log_to_file) {
+    if (write_json_to_file(jobj_root, target_file) == -1)
+      piksi_log(LOG_ERR, "Failed to write to file");
+  }
+}
+
+static int parse_options(int argc, char *argv[])
+{
+  enum { OPT_ID_METRICS_PATH = 1 };
+
+  const struct option long_opts[] = {
+    {"metrics_path", required_argument, 0, OPT_ID_METRICS_PATH},
+    {0, 0, 0, 0},
+  };
+
+  int opt;
+  while ((opt = getopt_long(argc, argv, "", long_opts, NULL)) != -1) {
+    switch (opt) {
+    case OPT_ID_METRICS_PATH: {
+      metrics_path = optarg;
+      break;
+    }
+    default: {
+      puts("Invalid option");
+      return -1;
+    }
+    }
+  }
+
+  return 0;
+}
+
+/**
+ * @brief used to trigger usage updates
+ */
+static void run_routine_function(pk_loop_t *loop, void *timer_handle, void *context)
+{
+  (void)loop;
+  (void)timer_handle;
+  (void)context;
+  write_metrics_to_file();
+}
+
+
+static int cleanup(int status)
+{
+  if (root_name != NULL) free(root_name);
+  logging_deinit();
+  sbp_deinit();
+  json_object_put(jobj_root);
```

Comment:
Since I don't know the json-c API it's a little unclear that this is decrementing a refcount.  Maybe we could assert that the refcount is 1 prior to calling this so that we're sure that it's actually getting freed?  A comment would help too.